web/src/main/java/org/cbioportal/web/util/DataBinHelper.java

```diff
@@ -332,4 +332,71 @@ public boolean isAgeAttribute(String attributeId) {
             }
         }
     }
+
+    public Set<BigDecimal> findDistinctValues(DataBin numericalBin, List<BigDecimal> numericalValues) {
+        Range<BigDecimal> range = calcRange(numericalBin);
+        
+        return numericalValues.stream().filter(range::contains).collect(Collectors.toSet());
+    }
+
+    public Set<Range<BigDecimal>> findDistinctSpecialRanges(DataBin numericalBin, List<Range<BigDecimal>> rangeValues) {
+        Range<BigDecimal> range = calcRange(numericalBin);
+        
+        return rangeValues.stream().filter(range::encloses).collect(Collectors.toSet());
+    }
+    
+    public List<DataBin> convertToDistinctBins(
+        List<DataBin> dataBins,
+        List<BigDecimal> numericalValues,
+        List<Range<BigDecimal>> rangeValues
+    ) {
+        List<DataBin> distinctBins = new ArrayList<>();
+        
+        for (DataBin bin: dataBins) {
+            Set<BigDecimal> distinctValues = this.findDistinctValues(bin, numericalValues);
+            Set<Range<BigDecimal>> distinctRanges = this.findDistinctSpecialRanges(bin, rangeValues);
+            
+            // if the bin contains only one distinct value and no range value then create a distinct bin
+            if (distinctRanges.size() == 0 && distinctValues.size() == 1 && this.areAllIntegers(distinctValues)) {
+                BigDecimal distinctValue = distinctValues.iterator().next();
+                
+                DataBin distinctBin = new DataBin();
+                distinctBin.setCount(bin.getCount());
+                distinctBin.setStart(distinctValue);
+                distinctBin.setEnd(distinctValue);
+                
+                distinctBins.add(distinctBin);
+            }
+            // else keep the bin as is
+            else {
+                distinctBins.add(bin);
+            }
+        }
+        
+        // all bins except the outlier bins has to be distinct,
+        // otherwise return the original input bins (no conversion)
+        if (areAllDistinctExceptOutliers(distinctBins)) {
```

Comment:
can't we just return `distinctBins`?

Reply:
good point. technically we can but we need to make sure that mixing distinct bins with internal bins doesn't mess up the bar chart layout in the frontend.